burst_detector/utils.py

```diff
@@ -36,29 +42,41 @@ def find_times_multi(
         cl_to_ind[clust_ids[i]] = i
 
     for i in range(sp_times.shape[0]):
-        if sp_clust[i] in cl_to_ind:
-            cl_times[cl_to_ind[sp_clust[i]]].append(sp_times[i])
+        time = sp_times[i]
+        if (
+            sp_clust[i] in cl_to_ind
+            and time >= pre_samples
+            and time < data.shape[0] - post_samples
+        ):
+            cl_times[cl_to_ind[sp_clust[i]]].append(time)
     for i in range(len(cl_times)):
         cl_times[i] = np.array(cl_times[i])
+        if max_spikes != -1 and cl_times[i].shape[0] > max_spikes:
+            # random shuffle and take the first max_spikes
+            np.random.shuffle(cl_times[i])
+            cl_times[i] = cl_times[i][:max_spikes]
 
     return cl_times
 
 
-def spikes_per_cluster(sp_clust: NDArray[np.int_]) -> NDArray[np.int_]:
+def spikes_per_cluster(sp_clust: NDArray[np.int_], max_spikes: int) -> NDArray[np.int_]:
     """
     Counts the number of spikes in each cluster.
 
-    ### Args
-        - `sp_clust` (nd.array): Spike cluster assignments.
+    Args
+        sp_clust (NDArray): Spike cluster assignments.
+        max_spikes (int): The maximum number of spikes to count for each cluster.
 
-    ### Returns
-        - `counts_array` (NDArray): Number of spikes per cluster, indexed by
+    Returns
+        counts_array (NDArray): Number of spikes per cluster, indexed by
             cluster id. Shape is (max_cluster_id + 1,).
 
     """
     ids, counts = np.unique(sp_clust, return_counts=True)
     counts_array = np.zeros(ids.max() + 1, dtype=int)
     counts_array[ids] = counts
+    if max_spikes != -1:
```

Comment:
we shouldn't be capping to max_spikes here, capping at max_spikes is just so we don't have performance issues for large clusters when we're doing extract_spikes or computing mean waveforms

Reply:
but if this doesn't affect any calculations then it's fine